quodlibet/quodlibet/ext/events/bansheeimport.py

```diff
@@ -0,0 +1,176 @@
+from quodlibet import app
+from quodlibet import ngettext
+from quodlibet import util
+from quodlibet.qltk.entry import UndoEntry
+from quodlibet.qltk import Icons
+from quodlibet.qltk.msg import Message, WarningMessage, ErrorMessage
+from quodlibet.util.path import expanduser, normalize_path
+from quodlibet.plugins.events import EventPlugin
+
+
+class BansheeDBImporter:
+
+    def __init__(self, library):
+
+        self._library = library
+        self._changed_songs = []
+
+    def read(self, db):
+        """Iterate through the database and import data for songs found in
+        the library
+        """
+
+        # use the Row class for extracting rows
+        db.row_factory = sqlite3.Row
+
+        # iterate over all songs in the database
+        # throws sqlite3.OperationalError if CoreTracks is not found
+        for row in db.execute("SELECT * FROM CoreTracks"):
+            try:
+                filename = uri2fsn(row["Uri"])
+            except ValueError:
+                continue
+
+            song = self._library.get(normalize_path(filename))
+            if not song:
+                continue
+
+            has_changed = False
+
+            # rating is stored as integer from 0 to 5
+            b_rating = row["Rating"] / 5.0
+            if b_rating != song("~#rating"):
+                song["~#rating"] = b_rating
+                has_changed = True
+
+            # play count is stored as integer from 0
+            if row["PlayCount"] != song("~#playcount"):
+                # summing play counts would break on multiple imports
+                song["~#playcount"] = row["PlayCount"]
+                has_changed = True
+
+            # skip count is stored as integer from 0
+            if row["SkipCount"] != song("~#skipcount"):
+                song["~#skipcount"] = row["SkipCount"]
+                has_changed = True
+
+            # timestamp is stored as integer or None
+            if row["LastPlayedStamp"] is not None:
+                value = row["LastPlayedStamp"]
+                # keep timestamp if it is newer than what we had
+                if value > song("~#lastplayed", 0):
+                    song["~#lastplayed"] = value
+                    has_changed = True
+
+            if row["DateAddedStamp"] is not None:
+                value = row["DateAddedStamp"]
+                # keep timestamp if it is older than what we had
+                if value < song("~#added", 0):
+                    song["~#added"] = value
+                    has_changed = True
+
+            if has_changed:
+                self._changed_songs.append(song)
+
+    def finish(self):
+        """Call at the end, also returns number of songs with data imported"""
+
+        count = len(self._changed_songs)
+        self._library.changed(self._changed_songs)
+        self._changed_songs = []
+        return count
+
+
+def do_import(parent, library):
+    db_path = expanduser(BansheeImport.USR_PATH)
+    db = sqlite3.connect(db_path)
```

Comment:
This can also raise OperationalError for weird paths ("/" for example), I'd move it into the try/except below.